File: src/main/java/rag/retrieval/KeywordRetriever.java
package main.java.rag.retrieval;

import main.java.rag.data.ChunkStore;
import main.java.rag.model.Chunk;
import main.java.rag.model.Hit;

import java.util.*;

public class KeywordRetriever implements Retriever {

    private final int defaultTopK;

    public KeywordRetriever() {
        this.defaultTopK = 10;
    }

    public KeywordRetriever(int topK) {
        this.defaultTopK = topK;
    }

    @Override
    public List<Hit> retrieve(List<String> queryTerms, ChunkStore store, int topK) {
        if (queryTerms == null || queryTerms.isEmpty()) return Collections.emptyList();
        int k = topK > 0 ? topK : defaultTopK;

        Map<String, Hit> hitMap = new HashMap<>();

        for (Chunk chunk : store.getAllChunks()) {
            String textLower = chunk.getText() == null ? "" : chunk.getText().toLowerCase();
            int tfSum = 0;

            for (String term : queryTerms) {
                if (term == null || term.isEmpty()) continue;
                tfSum += countOccurrences(textLower, term.toLowerCase());
            }

            if (tfSum > 0) {
                String key = chunk.getDocId() + "||" + chunk.getChunkId();
                hitMap.put(key, new Hit(chunk.getDocId(), chunk.getChunkId(), tfSum));
            }
        }

        List<Hit> hits = new ArrayList<>(hitMap.values());
        hits.sort(Comparator
                .comparingInt(Hit::getScore).reversed()
                .thenComparing(Hit::getDocId)
                .thenComparing(Hit::getChunkId));

        return k > 0 && hits.size() > k ? hits.subList(0, k) : hits;
    }

    private int countOccurrences(String text, String term) {
        int count = 0;
        int idx = text.indexOf(term);
        while (idx != -1) {
            count++;
            idx = text.indexOf(term, idx + term.length());
        }
        return count;
    }
}

File: iteration2/src/retrieval/keyword_retriever.py
from typing import Dict, List, Optional
import re

from src.data.chunk_store import ChunkStore
from src.model.hit import Hit
from src.retrieval.retriever import Retriever
from src.writer.simple_stemmer import SimpleStemmer


class KeywordRetriever(Retriever):

    def __init__(self, top_k: int, stemmer: Optional[SimpleStemmer] = None):
        self.__top_k = top_k
        self.__stemmer = stemmer

    def __safe_lower(self, s: str) -> str:
        return s.lower() if s is not None else ""

    def __extract_words(self, text: str) -> List[str]:
        words = re.findall(r'\b\w+\b', text.lower())
        if self.__stemmer:
            return [self.__stemmer.stem(word) for word in words]
        return words

    def __count_term_matches(self, chunk_words: List[str], query_term: str) -> int:
        if not chunk_words or not query_term:
            return 0
        
        query_term_lower = query_term.lower()
        query_stem = self.__stemmer.stem(query_term_lower) if self.__stemmer else query_term_lower
        
        return chunk_words.count(query_stem)

    def retrieve(self, query_terms: List[str], store: ChunkStore) -> List[Hit]:
        if not query_terms:
            return []

        hit_map: Dict[str, Hit] = {}

        for chunk in store.get_all_chunks():
            chunk_text = self.__safe_lower(chunk.get_text())
            chunk_words = self.__extract_words(chunk_text)
            
            total_tf = 0

            for term in query_terms:
                if not term:
                    continue
                total_tf += self.__count_term_matches(chunk_words, term)

            if total_tf <= 0:
                continue

            key = f"{chunk.get_doc_id()}||{chunk.get_chunk_id()}"
            hit = hit_map.get(key)

            if hit is None:
                hit = Hit(chunk.get_doc_id(), chunk.get_chunk_id(), total_tf)
                hit_map[key] = hit
            else:
                hit.set_score(hit.get_score() + total_tf)

        hits: List[Hit] = list(hit_map.values())

        hits.sort(key=lambda h: (
            -h.get_score(),           
            h.get_doc_id(),           
            h.get_chunk_id()
        ))
        
        return hits[:self.__top_k]
